internal/controller/controllers/agent_controller.go

```diff
@@ -348,6 +363,35 @@ func (r *AgentReconciler) populateEventsURL(log logrus.FieldLogger, agent *aiv1b
 	return nil
 }
 
+func generateControllerLogsDownloadURL(baseURL string, clusterID string, authType auth.AuthType, filterBy ...string) (string, error) {
+	path := fmt.Sprintf("%s%s/v2/clusters/%s/logs",
+		baseURL, restclient.DefaultBasePath, clusterID)
+	u, err := url.Parse(path)
+	if err != nil {
+		return "", err
+	}
+	queryParams := u.Query()
+	for i := 0; i < len(filterBy)-1; i += 2 {
+		queryParams.Set(filterBy[i], filterBy[i+1])
+	}
+	u.RawQuery = queryParams.Encode()
+	downloadURL := u.String()
```

Comment:
One option is to try to extract this into a separate function to share across events and logs, but I think this could be simplified quite a bit by using the generated url builder functions we already have access to along with simplifying parameters to this function (we don't support arbitrary filtering on logs only by host and type, so just pass those).

Then this can be simplified to something like:

```
import "github.com/openshift/assisted-service/restapi/operations/installer"
...
func generateControllerLogsDownloadURL(baseURL string, clusterID string, authType auth.AuthType, hostID *strfmt.UUID, logsType *string) (string, error) {
  builder := &installer.V2DownloadClusterLogsURL{
    ClusterID: strfmt.UUID(clusterID),
    HostID: hostID,
    LogsType: logsType,
  }

  u, err := builder.Build()
  if err != nil {
    return "", err
  }

  downloadURL := fmt.Sprintf("%s%s", baseURL, u.RequestURI())

  if authType != auth.TypeLocal {
    return downloadURL, nil
  }

  downloadURL, err = gencrypto.SignURL(downloadURL, clusterID, gencrypto.ClusterKey)
  if err != nil {
    return "", errors.Wrap(err, "failed to sign cluster controller logs URL")
  }

  return downloadURL, nil
}
```

This gets rid of all the repeated indexing and parsing logic.

Reply:
Thanks for letting me know about the URL builder! I've modified the code to use this, how does it look?